### PR TITLE
Add T5 backbone

### PR DESCRIPTION
**Pull request by fchollet**
T5 backbone implementation.

Some inner parts of the model are ported from https://github.com/huggingface/transformers/blob/main/src/transformers/models/t5/modeling_tf_t5.py

All checkpoints are correctly loadable.

### CODE REVIEW

**Comment by fchollet** [comment]
New: collapsed CrossAttention and SelfAttention classes into one. It could actually be collapsed further into the T5Attention classes, but that class is already a bit messy. Some more style fixes as well.

File: keras_nlp/models/t5/t5_backbone.py [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+        token_ids = keras.Input(shape=(None,), dtype="int32", name="token_ids")
+        decoder_token_ids = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_token_ids"
+        )
+        encoder_outputs = encoder(token_ids)
+        decoder_outputs = decoder(
+            decoder_token_ids,
+            encoder_hidden_states=encoder_outputs,
+        )
+        super().__init__(
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:88]: I think we should probably just follow this input and output for now...
https://github.com/keras-team/keras-nlp/blob/4c94a0d68872a72ed3e27d3f93a3d7eec3ddc0ba/keras_nlp/models/bart/bart_backbone.py#L217-L226

Padding mask vs whole attention mask is unclear to me, but our convention is to take a padding mask on every model, so let's do it here.

And it does seem useful for the backbone to return both the encoder and decoder outputs? That gives you the full "dense representation" at the backbone level.

File: keras_nlp/models/t5/t5_backbone.py [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+        layer_norm_epsilon,
+        **kwargs,
+    ):
+        shared_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=output_dim,
+            embeddings_initializer=keras.initializers.TruncatedNormal(1.0),
+            name="shared_embedding",
+        )
+        encoder = T5MainLayer(
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:51]: I think it would be a little more readable to consider removing "main layer" class entirely and switching to something like this...

```
encoder_embedding = token_embedding(encoder_token_ids)
decoder_embedding = token_embedding(decoder_token_ids)

# Set up masks.

for i in range(num_layers):
  # Call encoder blocks.

for in in range(num_layers):
  # Call decoder blocks.
```

There would be a few advantages: 
- There a lot of state passing that would be way easier to follow if there wasn't a shared encoder/decoder main class. E.g. all the "encoder state" `encoder_hidden_states`, `encoder_attention_mask`, `encoder_decoder_position_bias` and `encoder_layer_head_mask` that gets passed to the decoder.
- Model summary would look similar to other backbones.
- `get_layer` calls would work similar to other backbones.

The main thing the main layer seems to do is mask setup. We can just move that here and gradually use more utilities for it.

File: keras_nlp/models/t5/t5_backbone.py [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+                "activation": self.activation,
+                "dropout": self.dropout,
+                "layer_norm_epsilon": self.layer_norm_epsilon,
+            }
+        )
+        return config
+
+    @classproperty
+    def presets(cls):
+        return copy.deepcopy(backbone_presets)
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:125]: We have started directly exposing the `token_embedding` as a property as it is pretty common to need to use it directly. https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/bart/bart_backbone.py#L255-L257

File: keras_nlp/models/t5/t5_presets.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        "weights_hash": "TODO",
+        "vocabulary_url": "TODO",
+        "vocabulary_hash": "TODO",
+        "merges_url": "TODO",
+        "merges_hash": "TODO",
+    },
+    "t5_3b": {
+        "metadata": {
+            "description": (
+                "'3B' configuration of T5 (2.85 billion parameters)."
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_presets.py:110]: We will already show the parameter count in our table on keras.io, so might want to make the description more orthogonal. You can see our current descriptions here https://keras.io/api/keras_nlp/models/#backbone-presets

**Review by mattdangerw — CHANGES_REQUESTED** [review]

File: keras_nlp/models/t5/t5_backbone.py [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+class T5Backbone(Backbone):
+    def __init__(
+        self,
+        vocabulary_size,
+        output_dim,
+        hidden_dim,
+        key_value_dim,
+        num_blocks,
+        num_heads,
+        use_gated_activation,
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:39]: As it turns out, yes, the gated activation setup is the more up-to-date version and has been shown to work better (e.g. in the training experiments by Elie's team).

File: keras_nlp/models/t5/t5_backbone.py [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+            )
+
+    @staticmethod
+    def _relative_position_bucket(
+        relative_position, bidirectional=True, num_buckets=32, max_distance=128
+    ):
+        """Adapted from Mesh Tensorflow:
+
+        https://github.com/tensorflow/mesh/blob/0cb87fe07da627bf0b7e60475d59f95ed6b5be3d/mesh_tensorflow/transformer/transformer_layers.py#L593
+        Translate relative position to a bucket number for relative attention. The relative position is defined as
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:285]: Done.

File: keras_nlp/models/t5/t5_backbone.py [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+from keras_nlp.models.t5.t5_presets import backbone_presets
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.T5Backbone")
+class T5Backbone(Backbone):
+    def __init__(
+        self,
+        vocabulary_size,
+        output_dim,
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:34]: > Lastly "num_blocks" we call "num_layers". Again it's a place there is no perfect consistency, but here is BERT referring to a 12 block model as a 12 layer model 

I disagree with this because the concept of Layer is pretty well established in Keras and is a lot more atomic than this.

File: keras_nlp/models/t5/t5_backbone.py [hunk 8]
```diff
@@ -0,0 +1,846 @@
...
+    def call(
+        self,
+        hidden_states,
+        mask=None,
+        key_value_states=None,
+        position_bias=None,
+        past_key_value=None,
+        layer_head_mask=None,
+        query_length=None,
+        training=False,
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:364]: We have dropout layers -- seems cleaner to pass the `training` arg to them

File: keras_nlp/models/t5/t5_backbone.py [hunk 9]
```diff
@@ -0,0 +1,846 @@
...
+
+def shape_list(tensor):
+    dynamic = tf.shape(tensor)
+    if tensor.shape == tf.TensorShape(None):
+        return dynamic
+    static = tensor.shape.as_list()
+    return [dynamic[i] if s is None else s for i, s in enumerate(static)]
+
+
+class T5LayerNorm(keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:136]: Just using `LayerNormalization(epsilon=..., center=False)` does not preserve numerics (not even close). I have no. investigated further.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/models/t5/t5_backbone.py:136]: Sounds good! This seems totally fine, was just wondering what was different with our stock layer.

File: keras_nlp/models/t5/t5_backbone.py (hunk 2 above)
Comment (jbischof) [reply keras_nlp/models/t5/t5_backbone.py:51]: Another way of saying this is that this implementation should look like our BART model unless there's a good reason to deviate.

**Review by jbischof — COMMENTED** [review]
Thanks for the contribution @fchollet!

My overall reaction is that this code really does look like a port from somewhere else and doesn't seem to follow many conventions of our repo. I would strongly prefer using BART as a starting point and adding some extra functionality as needed so our repo feels like a consistent product.

File: keras_nlp/models/t5/t5_backbone.py [hunk 10]
```diff
@@ -0,0 +1,846 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.token_embeddings = token_embeddings
+        self.is_decoder = is_decoder
+        self.num_hidden_layers = num_blocks
+
+        self.blocks = [
+            T5Block(
+                is_decoder,
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:680]: Done.

File: keras_nlp/models/t5/t5_backbone.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:34]: I applied the other name substitutions and remove `key_value_dim`.

File: keras_nlp/models/t5/t5_backbone.py [hunk 11]
```diff
@@ -0,0 +1,846 @@
...
+                f"or `{err_msg_prefix}inputs_embeddings`."
+            )
+
+        if inputs_embeddings is None:
+            inputs_embeddings = self.token_embeddings(token_ids)
+
+        batch_size, seq_length = input_shape
+
+        # required mask seq length can be calculated via length of past
+        mask_seq_length = (
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:734]: SG, let's do this in a follow up.

File: keras_nlp/models/t5/t5_backbone.py (hunk 3 above)
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:125]: Done.

File: keras_nlp/models/t5/t5_presets.py (hunk 4 above)
Comment (fchollet) [reply keras_nlp/models/t5/t5_presets.py]: There isn't that much useful info we can bring up in the one-line description -- I included the number of layers and a more user-readable parameter count, which is ultimately the defining characteristic of each checkpoint.

File: keras_nlp/models/t5/t5_backbone.py [hunk 12]
```diff
@@ -0,0 +1,846 @@
...
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.models.backbone import Backbone
+from keras_nlp.models.t5.t5_presets import backbone_presets
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.T5Backbone")
+class T5Backbone(Backbone):
+    def __init__(
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py:81]: Added a docstring -- still needs code examples though

File: keras_nlp/models/t5/t5_backbone.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+    T5 was introduced in
+    [Exploring the Limits of Transfer Learning with a Unified Text-to-Text Transformer](https://arxiv.org/abs/1910.10683)
+
+    The default constructor gives a fully customizable, randomly initialized T5
+    model with any number of layers, heads, and embedding dimensions. To load
+    preset architectures and weights, use the `from_preset` constructor.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:50]: I think we decided that for google models we don't need to link out (as long as mit or apache). Just stop at the first sentence.

File: keras_nlp/models/t5/t5_backbone.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+        dropout: float. Dropout probability for the Transformer layers.
+        activation: activation function (or activation string name). The
+            activation to be used in the inner dense blocks of the
+            Transformer layers. Defaults to `"gelu"`. The original
+            T5 architecture used `"relu"`,
+            but the modern version uses `"gelu"`.
+        use_gated_activation: boolean. Whether to use activation gating in
+            the inner dense blocks of the Transformer layers.
+            Defaults to True. The original T5 architecture didn't use
+            gating, but the version does.
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:71]: but the version does -> but more recent versions do\

We could also link to https://arxiv.org/abs/2002.05202 , but maybe overkill

File: keras_nlp/models/t5/t5_backbone.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+        layer_norm_epsilon: Epsilon factor to be used in the
```
Comment (mattdangerw) [inline keras_nlp/models/t5/t5_backbone.py:72]: add "float."

**Review by mattdangerw — APPROVED** [review]
LGTM!

File: keras_nlp/models/t5/t5_backbone.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+            layer normalization layers in the Transformer layers.
+
+    Examples:
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py]: Done.

File: keras_nlp/models/t5/t5_backbone.py (hunk 15 above)
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py]: Done.

File: keras_nlp/models/t5/t5_backbone.py (hunk 14 above)
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py]: Done.

File: keras_nlp/models/t5/t5_backbone.py (hunk 13 above)
Comment (fchollet) [reply keras_nlp/models/t5/t5_backbone.py]: Done.

**Comment by fchollet** [comment]
Addressed all comments!
